cpp/src/arrow/array/array_test.cc

```diff
@@ -78,6 +78,24 @@ class TestArray : public ::testing::Test {
   MemoryPool* pool_;
 };
 
+TEST_F(TestArray, ValidateFullNullableList) {
+  auto f1 = field("f1", int32(), /*nullable=*/false);
+  auto ty = list(f1);
+  auto array = ArrayFromJSON(ty, "[[0, 1, 2, null], null, [4, 5]]");
+  ASSERT_RAISES(Invalid, array->ValidateFull());
```

Comment:
Can we split the checks between nested nulls and toplevel nulls just like below?

Reply:
The check for the top level fails. Will see if it is due to "ArrayFromJson"